Node collapse

kwcantrell opened the pull request:
I am creating this PR to jump start the review process since this PR is fairly large. I am still working on finishing up some tests however, the important parts are tested.

 
Notes
----------------
Each layout has a different method to collapse clades as seen below:
| Unrooted | Rectangular | Circular|
| ------------ | ---------------- | ---------- |
| ![collapse-unroot](https://user-images.githubusercontent.com/28127694/88436012-41cf4280-cdb8-11ea-9294-05bcbbaf6f64.png) |  ![collapse-rectangular](https://user-images.githubusercontent.com/28127694/88436035-4bf14100-cdb8-11ea-9b5b-b41163e664b9.png) | ![collapse-circ](https://user-images.githubusercontent.com/28127694/88436053-53184f00-cdb8-11ea-94c7-0259b0635844.png)

In addition, Rectangular and Circular layout will also have a "symmetric" version as seen below:
| Normal | Symmetric |
| ---------- | -------------- |
| ![collapse-rectangular](https://user-images.githubusercontent.com/28127694/88436035-4bf14100-cdb8-11ea-9b5b-b41163e664b9.png) | ![rec-sym](https://user-images.githubusercontent.com/28127694/88436231-b99d6d00-cdb8-11ea-904f-448c730fbd4d.png) |
| ![collapse-circ](https://user-images.githubusercontent.com/28127694/88436053-53184f00-cdb8-11ea-94c7-0259b0635844.png) |  ![cir-sym](https://user-images.githubusercontent.com/28127694/88436265-cb7f1000-cdb8-11ea-887b-7107e2c2be89.png) |

Collapsed clade can also be clicked on. Doing so, at the time being, is equivalent to selecting the root node. In the future, clicking on a clade will display information about the clade (such as its group membership, # tips, # internal node,...)

It should also be noted that an issue can occur with the Unrooted/Circular layout if the clade is more than a 180°:
![circ-bug](https://user-images.githubusercontent.com/28127694/88436689-a212b400-cdb9-11ea-85cd-2b6cd8b9eb23.png)
![unroot-lay](https://user-images.githubusercontent.com/28127694/88436695-a4750e00-cdb9-11ea-843f-a0f32cdc948c.png)

This is not a bug per se but a consequence of the collapsing algorithm (I believe iTol shares the same issue). This happens because the "left" and "right" most children of the clade are behind the root. This, in turn, causes the clade shape to do one of two things: point in the wrong direction (this happened in the above Unrooted layout) or overlap other clades/root (this happened in the circular layout).


Minor changes:
- cangle has been added treeData for each tip
- lengths has been added to bp-tree
- an in-order tree traversal has been added to bp-tree (and will be cached after it is called)
- inSampe has been removed from treeData
- hide checkbox has been replaced with collapse checkbox
- setNonSampleBranchVisibility() has been removed

ElDeveloper commented on empress/support_files/js/empress.js:
```diff
@@ -1230,5 +1264,621 @@ define([
+            // triangle from 1, 4, 3
+            addTriangle(cladeInfo.right);
+        } else if (this._currentLayout === "Rectangular") {
+            // Rectangular layou is a triangle. Symmetric version is used if
```
```suggestion
            // Rectangular layout is a triangle. Symmetric version is used if
```

ElDeveloper commented on empress/support_files/js/empress.js:
```diff
@@ -1230,5 +1264,621 @@ define([
+        this._collapsedCladeBuffer.push(...cladeBuffer);
+    };
+
+    /**
```
MInor but this docstring is probably missing the `@private` tag.

ElDeveloper commented on empress/support_files/js/empress.js:
```diff
@@ -1230,5 +1264,621 @@ define([
+
+        // find left most child
+        // Note: initializing lchild as cladeRoot incase cladeRoot is a tip
+        lchild = cladeRoot;
```
```suggestion
        var lchild = cladeRoot;
```

ElDeveloper commented on empress/support_files/js/empress.js:
```diff
@@ -1230,5 +1264,621 @@ define([
+            // clade, we first calculate the area of the collapsed clade.
+            // (The shape of the collapsed clade is a quad whose vertices are
+            // (1) root, (2) "left" most child, (3) "right" most child, and
+            // (4) "deepest" child). Next, we form four triangls whose vertices
```
```suggestion
            // (4) "deepest" child). Next, we form four triangles whose vertices
```

ElDeveloper commented on empress/support_files/js/empress.js:
```diff
@@ -1230,5 +1264,621 @@ define([
+     *                  the collapse clade will be returned otherwise -1 is
+     *                  returned.
+     */
+    Empress.prototype.isInAClade = function (point) {
```
The name of this method is a bit misleading, would likely be more appropriate as `getRootNodeForPointInClade` - or something like that up to you. `isInClade` makes it sound like it returns a boolean value if the point is in a clade, but really it is returning the root for a clade if the point falls in that space.

ElDeveloper commented on tests/test-empress.js:
```diff
@@ -677,5 +685,115 @@ require([
         });
+
+        test("Test collapseClades", function () {
+            // red should be the only collapsable clade
```
```suggestion
            // red should be the only collapsible clade
```

ElDeveloper commented on empress/support_files/templates/side-panel.html:
```diff
@@ -174,6 +188,28 @@
+  </p>
+  <p>
+    <label for="symmetric">Symmetric</label>
+    <input type="radio" id="symmetric" name="collapse-radio" value="symmetric"
```
It would be nice if this input could be disabled when the unrooted layout is set.

ElDeveloper commented on empress/support_files/templates/side-panel.html:
```diff
@@ -132,10 +143,13 @@
-    <label for="animate-hide-non-feature">Hide Non-Sample Features</label>
-    <input id="animate-hide-non-feature" type="checkbox" value="false"
+    <label for="animate-collapse-chk">Collapse Clades</label>
+    <input id="animate-collapse-chk" type="checkbox" value="false"
```
Should this input be disabled once the animation starts?

kwcantrell commented on empress/support_files/js/empress.js:
```diff
@@ -1230,5 +1264,621 @@ define([
+                group[node] !== -1 &&
+                color !== this.DEFAULT_COLOR
+            ) {
+                this._collapseClade(node);
```
Good catch! It would definitely collapse two clades that have the same color but different metadata values. Likely, should be an easy fix. Instead of using the color to group nodes, I can just use metadata.

kwcantrell commented:
@ElDeveloper thank you for the comments. I believe I have addressed all of your comments. Note that clades are now collapsed based on the _group array (which uses metadata value to set the groups) instead of color. I also added some more tests. 

Also, I don't have an issue with moving the collapse code to a different module. Although I think we should push the to another PR since it would require a bit of refactoring.

ElDeveloper commented on empress/support_files/js/empress.js:
```diff
@@ -1502,6 +1561,619 @@ define([
+        // Note: if the root of a clade has DEFUALT_COLOR then it will not be
+        // collapsed (since all of its children will also have DEFAULT_COLOR)
+        var inorder = this._tree.inOrderNodes();
+        for (var node in inorder) {
+            node = inorder[node];
```
This pattern can be changed for:

_.each(inorder, function(node) {
    
})

ElDeveloper commented on tests/python/test_core.py:
```diff
@@ -254,6 +254,7 @@ def test_to_dict(self):
         viz = Empress(self.tree, self.table, self.sample_metadata,
                       filter_unobserved_features_from_phylogeny=False)
         obs = viz._to_dict()
+        print(obs)
```
```suggestion
```

ElDeveloper commented on tests/test-animation-panel-handler.js:
```diff
@@ -158,34 +158,60 @@ require([
         this.panel.animator.totalFrames = 3;
-        this.panel.animator.curFrame = 1;
+
+        // Note: this use to be 1 but the animator class has been updated
```
```suggestion
        // Note: this used to be 1 but the animator class has been updated
```

ElDeveloper commented:
Thanks @kwcantrell
